src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores-export-button.component.ts

```diff
@@ -0,0 +1,116 @@
+                        );
+                        // Nest submission into participation so that it is available for the result component
+                        if (result.participation && result.submission) {
+                            result.participation.submissions = [result.submission];
+                        }
+                        return result;
+                    });
+                }),
+            );
+    }
+
+    /**
+     * Exports the exercise results as a csv file
+     */
+    exportResults() {
+        if (this.exercises.length === 0 && this.exercise !== undefined) {
+            this.exercises = this.exercises.concat(this.exercise);
+        }
+        this.exercises.forEach((exercise) => {
+            this.exercise = exercise;
+            this.getResults().subscribe((data) => {
+                const rows: string[] = [];
+                const results = data.body || [];
+                if (results.length === 0) {
+                    this.jhiAlertService.warning(`artemisApp.exercise.exportResultsEmptyError`, { exercise: exercise.title });
+                    window.scroll(0, 0);
+                    return;
+                }
+                results.forEach((result, index) => {
+                    const studentParticipation = result.participation! as StudentParticipation;
+                    const { participantName, participantIdentifier } = studentParticipation;
+                    const score = round(result.score);
+
+                    if (index === 0) {
+                        const nameAndUserNameColumnHeaders = studentParticipation.team ? 'Team Name,Team Short Name' : 'Name,Username';
+                        const optionalStudentsColumnHeader = studentParticipation.team ? ',Students' : '';
+                        if (this.exercise.type !== ExerciseType.PROGRAMMING) {
+                            rows.push(`data:text/csv;charset=utf-8,${nameAndUserNameColumnHeaders},Score${optionalStudentsColumnHeader}`);
+                        } else {
+                            rows.push(`data:text/csv;charset=utf-8,${nameAndUserNameColumnHeaders},Score,Repo Link${optionalStudentsColumnHeader}`);
+                        }
+                    }
+                    const optionalStudentsColumnValue = studentParticipation.team ? `,"${studentParticipation.team?.students?.map((s) => s.name).join(', ')}"` : '';
+                    if (this.exercise.type !== ExerciseType.PROGRAMMING) {
+                        rows.push(`${participantName},${participantIdentifier},${score}${optionalStudentsColumnValue}`);
+                    } else {
+                        const repoLink = (studentParticipation as ProgrammingExerciseStudentParticipation).repositoryUrl;
+                        rows.push(`${participantName},${participantIdentifier},${score},${repoLink}${optionalStudentsColumnValue}`);
+                    }
+                });
+                const csvContent = rows.join('\n');
```

Comment:
I guess some code here is duplicated now. Could you try to extract functions for the parts that include duplicated code?

Reply:
With https://github.com/ls1intum/Artemis/pull/3263/commits/1137c9a6defe09072d085ba33e8f05c561f265e6 I extracted the download logic (duplicate in `exercise-scores.component.ts - exportNames()`) but couldn't find any further part duplicating the same functionality of `exportNames()`